### ISSUE TITLE
How do you know when delay is finished?

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#180]
Need to bypass FX... well more specifically, the oscilloscope takes 1.6% cpu idle (in debug mode, but still...). Is there a way to tell when the buffer is basically silent/empty, maybe below 1.e-6 amplitude?

**Comment by elanhickler** [comment]
also, how do you get rid of this: 
![image](https://user-images.githubusercontent.com/9116575/35258689-6246c472-ffbe-11e7-80df-52bd09efc045.png)

**Comment by RobinSchmidt** [comment]
strictly speaking, it will have finished only when the signal falls soo low that it will be flushed to zero (by our denormal flush-to-zero code) - which can take a really long time. for practical matters, i'd probably use a threshold of something like -100dB - but obviously, the choice of that number is a matter of judgement. 1.e-6 - one part in a million - i think that's -60 dB - it's a value typically used to define reverb time (RT60 is a common measure for the length of f reverb tail). . soo - yeah - pick something in that ballpark

**Comment by RobinSchmidt** [comment]
oh - these pixel values is in ColourSchemeComponent.cpp - you must set a boolean value to false in paintOverChildren. i use that in debug mode to figure out optimal editor sizes

**Comment by RobinSchmidt** [comment]
unfortunately, when the grid-lines look homogenous, they are two pixels wide (and have half the brightness). maybe i should offset all pixel coordinates by 0.5. must be something about what you consider the pixel coodinates to be - top-left or center of the pixel - and my drawing code makes a wrong assumption of the convention used by juce

**Comment by RobinSchmidt** [comment]
...i think, the rule is: if there are N grid-lines, your pixel-width (or height) of the plot should be divisible by N+1 (because there are N+1 sections...or something...just some gut math, i didn't consult pencil and paper)

**Comment by elanhickler** [comment]
what are you talking about? grid lines? offset by .5? your drawing code makes wrong assumption?

**Comment by RobinSchmidt** [comment]
yes, you are right. -60dB is 0.001. so, yes, -120dB seems appropriate to me. if you don't want to scan the entire buffer, you could use a formula based on the delaytime and the feedback gain to compute the time, after which the output signal has decayed to -120dB and switch to bypass after that time has passed. then you would need to count samples, though

**Comment by RobinSchmidt** [comment]
> what are you talking about? grid lines? offset by .5? your drawing code makes wrong assumption?

yes, try a pixel size for a eq-plot for which the grid lines of the plot appear homogenous. then, they are 2 pixels thick with 0.5 for the brightness. better would be 1 pixel thick at full brightness. ...i'll probably rewrite my coordinate-system drawing code anyway. it's a huge mess with lots of code duplication (very old code)

**Comment by RobinSchmidt** [comment]
> use a formula 

...i think, you need to start with `a(t) = k^(t/d)` where t: time (after input went silent,) a(t):  amplitude at time t, k: feedback gain, d: delaytime, so the formula for the time t should be `t = d * log(a)/log(k)`, if i'm not mistaken
